src/Avalonia.Controls/Flyouts/FlyoutBase.cs

```diff
@@ -0,0 +1,487 @@
+﻿using System;
+using System.ComponentModel;
+using Avalonia.Input;
+using Avalonia.Input.Raw;
+using Avalonia.Layout;
+using Avalonia.Logging;
+
+#nullable enable
+
+namespace Avalonia.Controls.Primitives
+{
+    public abstract class FlyoutBase : AvaloniaObject
+    {
+        static FlyoutBase()
+        {
+            Control.ContextFlyoutProperty.Changed.Subscribe(OnContextFlyoutPropertyChanged);
+        }
+
+        /// <summary>
+        /// Defines the <see cref="IsOpen"/> property
+        /// </summary>
+        private static readonly DirectProperty<FlyoutBase, bool> IsOpenProperty =
+           AvaloniaProperty.RegisterDirect<FlyoutBase, bool>(nameof(IsOpen),
+               x => x.IsOpen);
+
+        /// <summary>
+        /// Defines the <see cref="Target"/> property
+        /// </summary>
+        public static readonly DirectProperty<FlyoutBase, Control?> TargetProperty =
+            AvaloniaProperty.RegisterDirect<FlyoutBase, Control?>(nameof(Target), x => x.Target);
+
+        /// <summary>
+        /// Defines the <see cref="Placement"/> property
+        /// </summary>
+        public static readonly StyledProperty<FlyoutPlacementMode> PlacementProperty =
+            AvaloniaProperty.Register<FlyoutBase, FlyoutPlacementMode>(nameof(Placement));
+
+        /// <summary>
+        /// Defines the <see cref="ShowMode"/> property
+        /// </summary>
+        public static readonly DirectProperty<FlyoutBase, FlyoutShowMode> ShowModeProperty =
+            AvaloniaProperty.RegisterDirect<FlyoutBase, FlyoutShowMode>(nameof(ShowMode),
+                x => x.ShowMode, (x, v) => x.ShowMode = v);
+
+        /// <summary>
+        /// Defines the AttachedFlyout property
+        /// </summary>
+        public static readonly AttachedProperty<FlyoutBase?> AttachedFlyoutProperty =
+            AvaloniaProperty.RegisterAttached<FlyoutBase, Control, FlyoutBase?>("AttachedFlyout", null);
+
+        private bool _isOpen;
+        private Control? _target;
+        protected Popup? _popup;
```

Comment:
Perhaps this should be a property instead? I haven't seen many (or any) protected fields in this codebase.

Reply:
Fixed